storm-core/src/jvm/org/apache/storm/metric/MetricsConsumerBolt.java

```diff
@@ -17,24 +17,45 @@
  */
 package org.apache.storm.metric;
 
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
 import org.apache.storm.Config;
 import org.apache.storm.metric.api.IMetricsConsumer;
 import org.apache.storm.task.IBolt;
 import org.apache.storm.task.OutputCollector;
 import org.apache.storm.task.TopologyContext;
 import org.apache.storm.tuple.Tuple;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.util.Collection;
+import java.util.List;
 import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
 
 public class MetricsConsumerBolt implements IBolt {
+    public static final Logger LOG = LoggerFactory.getLogger(MetricsConsumerBolt.class);
+
     IMetricsConsumer _metricsConsumer;
     String _consumerClassName;
     OutputCollector _collector;
     Object _registrationArgument;
+    private final int _maxRetainMetricTuples;
+    private Predicate<IMetricsConsumer.DataPoint> _filterPredicate;
+
+    private final BlockingQueue<MetricsTask> _taskQueue = new LinkedBlockingDeque<>();
+    private Thread _taskExecuteThread;
+    private volatile boolean _running = true;
+
+    public MetricsConsumerBolt(String consumerClassName, Object registrationArgument, int maxRetainMetricTuples,
+                               Predicate<IMetricsConsumer.DataPoint> filterPredicate) {
 
-    public MetricsConsumerBolt(String consumerClassName, Object registrationArgument) {
         _consumerClassName = consumerClassName;
         _registrationArgument = registrationArgument;
+        _maxRetainMetricTuples = maxRetainMetricTuples;
```

Comment:
OK. We could even just set queue unbounded when it's lower than 0. I'll address it.